### PR TITLE
Support an allowlist to manage available options for 'state-into-spec'

### PR DESCRIPTION
**Pull request by maqiuyujoyce**
### Change description

Fixes b/328153703

* Added an allowlist including all the existing v1beta1 kinds. The allowlist is used to determine whether the GVK should support `state-into-spec: merge`.
* Add the logic in the defaulting and validation for the state-into-spec annotation to avoid any newly added v1beta1 resources to use `state-into-spec: merge`.

### Tests you have done

- [N/A] Run `make ready-pr` to ensure this PR is ready for review.
- [X] Perform necessary E2E testing for changed resources.

Downloaded #1271, rebased on top of this PR, then ran the following test command:

```
go test -timeout 1200s -v -tags=integration ./pkg/controller/dynamic/ -test.run TestCreateNoChangeUpdateDelete -run-tests vertexaidatasetbasic
```

The created resource has `state-into-spec: absent`.

### CODE REVIEW

**Comment by maqiuyujoyce** [comment]
This should unblock #1281 and should partly unblock #1271.

File: pkg/k8s/state_into_spec_annotation.go [hunk 1]
```diff
@@ -16,34 +16,50 @@ package k8s
...
 
-	if !isAcceptedValue(val, StateIntoSpecAnnotationValues) {
-		return fmt.Errorf("invalid value '%v' for '%v' annotation, can be one of {%v}", val, StateIntoSpecAnnotation, strings.Join(StateIntoSpecAnnotationValues, ", "))
+	if !isAcceptedStateIntoSpecValue(val, obj.GetObjectKind().GroupVersionKind()) {
+		return fmt.Errorf("invalid value '%v' for '%v' annotation in kind %v", val, StateIntoSpecAnnotation, obj.GetObjectKind().GroupVersionKind().Kind)
 	}
 	return nil
 }
 
+func setStateIntoSpecDefaultValue(defaultValue string, obj client.Object) {
```
[the remark this answers is not in the document]
Comment (maqiuyujoyce) [reply pkg/k8s/state_into_spec_annotation.go:47]: What if the function name is changed to `setStateIntoSpecDefaultValueIfAllowed` or the param name is changed to something like `candidateDefaultValue`?

[the remark this answers is not in the document]
Comment (jingyih) [reply pkg/k8s/state_into_spec_annotation.go:47]: `setStateIntoSpecDefaultValueIfAllowed` sounds good to me.

**Pull request merged**